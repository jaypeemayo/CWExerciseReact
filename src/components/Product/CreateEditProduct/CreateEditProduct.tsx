import React from "react";
import Modal from "react-modal";
import { ICreateEditProduct } from ".";
import { IProduct } from "..";
import 'bootstrap/dist/css/bootstrap.min.css';

export const CreateEditProduct: React.FunctionComponent<ICreateEditProduct> = ({ OnSavedProduct, OnCloseCreateUpdate, productToEdit }) => {
  const initialiseProduct = () => {
    if (productToEdit) {
      return productToEdit;
    } else {
      return { name: "", price: 0, type: 1, active: false } as IProduct;
    }
  };
  const [product, setProduct] = React.useState<IProduct>(initialiseProduct());
  const handleNameChange = (e) => {
    setProduct({
      ...product,
      name: e.target.value
    });
  };
  const handleNumberChange = (e) => {
    var x = Number(e.target.value).toFixed(2);
    setProduct({
      ...product,
      price: Number(x)
    });
  };
  const handleTypeChange = (e) => {
    setProduct({
      ...product,
      type: e.target.value
    });
  };
  const handleActiveChange = (e) => {
    setProduct({
      ...product,
      active: e.target.checked
    });
  };

  const handleSave = () => {
    OnSavedProduct(
      productToEdit === null
        ? true
        : false,
      product);
  };

  const customStyles = {
    content: {
      top: "50%",
      left: "50%",
      right: "auto",
      bottom: "auto",
      marginRight: "-50%",
      transform: "translate(-50%, -50%)",
      padding: "0px",
      width: "400px",
    }
  };

  return (
    <Modal ariaHideApp={false} isOpen={true} style={customStyles}>
      <div className="modal-content">
        <div className="modal-header">
          <h5>
            {
              productToEdit === null
                ? "Create Product"
                : "Update Product"
            }
          </h5>
        </div>
        <div className="modal-body">
          <form>
            <div className="form-group pb-3">
              <label htmlFor="name">Name:</label>{" "}
              <input className="form-control" defaultValue={product.name} onChange={handleNameChange} maxLength={10} type="text" name="name" id="name" />
            </div>

            <div className="form-group pb-3">
              <label htmlFor="price">Price:</label>{" "}
              <input className="form-control" onChange={handleNumberChange} value={product.price} type="number" name="price" id="price" step='0.01' placeholder='0.00' />
            </div>

            <div className="form-group pb-3">
              <label id="type">Type</label>
              <select className="form-control" value={product.type} onChange={handleTypeChange}>
                <option value="1">Books</option>
                <option value="2">Electronics</option>
                <option value="3">Food</option>
                <option value="4">Furniture</option>
                <option value="5">Toys</option>
              </select>
            </div>

            <div className="form-check">
              <input className="form-check-input" defaultChecked={product.active} onChange={handleActiveChange} type="checkbox" name="active" id="active" />
              <label htmlFor="active">Active</label>
            </div>
          </form>
        </div>

        <div className="modal-footer">
          <button className="btn btn-secondary" onClick={OnCloseCreateUpdate}>Close</button>
          <button className="btn btn-primary" onClick={handleSave}>Save</button>
        </div>
      </div>
    </Modal>
  );
};
