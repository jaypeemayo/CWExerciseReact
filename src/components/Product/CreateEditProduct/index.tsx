import { IProduct } from "..";

export interface ICreateEditProduct {
  OnSavedProduct(isCreate: boolean, product: IProduct): void;
  OnCloseCreateUpdate(): void;
  productToEdit?: IProduct;
}
