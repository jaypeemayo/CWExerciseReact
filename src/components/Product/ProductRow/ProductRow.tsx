import React from "react";
import { IProductRow } from ".";
import { FaEdit, FaTrash } from "react-icons/fa";
import "bootstrap/dist/css/bootstrap.min.css";

export const ProductRow: React.FunctionComponent<IProductRow> = ({
  productID,
  name,
  price,
  type,
  active,
  OnDelete,
  OnEdit
}) => {
  return (<tr>
    <td>{productID}</td>
    <td>{name}</td>
    <td>{price}</td>
    <td>{type}</td>
    <td>{
      active
        ? "true"
        : "false"
    }</td>
    <td>
      <a className="p-4" title="Delete" href="javascript:void(0)" onClick={() => {
        OnDelete(productID);
      }}>
        <FaTrash />
      </a>
      <a title="Delete" href="javascript:void(0)" onClick={() => {
        OnEdit({ productID, name, price, type, active });
      }}>
        <FaEdit />
      </a>
    </td>
  </tr>);
};
