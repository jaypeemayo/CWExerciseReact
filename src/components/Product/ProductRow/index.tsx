import { IProduct } from "..";

export interface IProductRow extends IProduct {
    OnDelete(productID: number): void;
    OnEdit(product: IProduct): void;
  }