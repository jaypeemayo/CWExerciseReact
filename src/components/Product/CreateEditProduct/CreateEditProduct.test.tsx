import { render } from "@testing-library/react";
import * as React from "react";
import { IProductRow } from "../ProductRow";
import { CreateEditProduct } from "./CreateEditProduct";


jest.mock('react-modal', () => {
    const React = require('react');
    const TestReactModal = require('../../../../tests/__mocks__/mockReactModal');
    return TestReactModal.default;
  });

test("Test snapshot", () => {
    const onSavedProduct = jest.fn();
    const onCloseCreateUpdate = jest.fn();
    var product = {
        productID: 1,
        name: "string",
        price: 2,
        type: 2,
        active: false,
    } as IProductRow;
    const {container, unmount} = render(
        <CreateEditProduct OnSavedProduct={onSavedProduct}  OnCloseCreateUpdate={onCloseCreateUpdate} productToEdit={product}>
        </CreateEditProduct>
        ,
    );

    expect(container).toMatchSnapshot();
    unmount();
});


