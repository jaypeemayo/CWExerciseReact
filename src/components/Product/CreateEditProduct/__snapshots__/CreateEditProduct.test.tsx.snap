// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot 1`] = `
<div>
  <div>
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <h5>
          Update Product
        </h5>
      </div>
      <div
        class="modal-body"
      >
        <form>
          <div
            class="form-group pb-3"
          >
            <label
              for="name"
            >
              Name:
            </label>
             
            <input
              class="form-control"
              id="name"
              maxlength="10"
              name="name"
              type="text"
              value="string"
            />
          </div>
          <div
            class="form-group pb-3"
          >
            <label
              for="price"
            >
              Price:
            </label>
             
            <input
              class="form-control"
              id="price"
              name="price"
              placeholder="0.00"
              step="0.01"
              type="number"
              value="2"
            />
          </div>
          <div
            class="form-group pb-3"
          >
            <label
              id="type"
            >
              Type
            </label>
            <select
              class="form-control"
            >
              <option
                value="1"
              >
                Books
              </option>
              <option
                value="2"
              >
                Electronics
              </option>
              <option
                value="3"
              >
                Food
              </option>
              <option
                value="4"
              >
                Furniture
              </option>
              <option
                value="5"
              >
                Toys
              </option>
            </select>
          </div>
          <div
            class="form-check"
          >
            <input
              class="form-check-input"
              id="active"
              name="active"
              type="checkbox"
            />
            <label
              for="active"
            >
              Active
            </label>
          </div>
        </form>
      </div>
      <div
        class="modal-footer"
      >
        <button
          class="btn btn-secondary"
        >
          Close
        </button>
        <button
          class="btn btn-primary"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;
